Update validators and procedures docs (#4245)
trpc/trpc: www/docs/server/validators.md

        name: z.string(),
      }),
    )
    .output(
      z.object({
        greeting: z.string(),
      }),
    )
    .query(({ input }) => {
      //      ^?
      return {
        greeting: `hello ${input.name}`,
      };
    }),
});

export type AppRouter = typeof appRouter;
```

### Writing a custom validator

If you like, you can simply pass a function as a validator, though this isn't necessarily the best choice to use widely unless you have a specific need.

```ts twoslash
import { initTRPC } from '@trpc/server';

export const t = initTRPC.create();

export const appRouter = t.router({
  hello: t.procedure
    .input((value: any): { name: string } => {
      if (value && typeof value.name === 'string') {
        return { name: value.name };
      }
      throw new Error('Greeting not found');
    })
    .output((value: any) => {
      if (value && typeof value.greeting === 'string') {
        return { greeting: value.greeting };
      }
      throw new Error('Greeting not found');
    })
    .query(({ input }) => {
      //      ^?
      return {
        greeting: `hello ${input.name}`,
      };
    }),
});

export type AppRouter = typeof appRouter;
```

> ### With [Yup](https://github.com/jquense/yup)

Reviewer: These could be tabs like they are on Quickstart?

Author: Oh I had a look at this and felt that actually the way things are is quite nice. Having a lot of linkable sections in the sidebar should pay us back later in visibility and linkability.

Author: Plus the list if quite long and may not scale well to tabs as it continues to grow

Validators are used to infer the input and output types of your procedures.

:::info
If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
:::

## Using a function as a validator

If you like, you can simply pass a function as a validator, though this isn't necessarily the best choice to use widely unless you have a specific need.

In most cases we recommend you use a [Validator](#validator-integrations)

```ts twoslash
import { initTRPC } from '@trpc/server';

export const t = initTRPC.create();

export const appRouter = t.router({
  hello: t.procedure
    .input((value: any): { name: string } => {
      if (value && typeof value.name === 'string') {
        return { name: value.name };
      }
      throw new Error('Greeting not found');
    })
    .output((value: any) => {
      if (value && typeof value.greeting === 'string') {
        return { greeting: value.greeting };
      }
      throw new Error('Greeting not found');
    })
    .query(({ input }) => {
      //      ^?
      return {
        greeting: `hello ${input.name}`,
      };
    }),
});

export type AppRouter = typeof appRouter;
```

> ## Validator integrations

Reviewer: tabs would be nice to compact this down

Author: C&P from the old thread as I guess it got lost:

I had a look at this and felt that actually the way things are is quite nice. Having a lot of linkable sections in the sidebar should pay us back later in visibility and linkability. Plus the list if quite long and may not scale well to tabs as it continues to grow

Author: This is quite pleasant in the sidebar

<img width="301" alt="image" src="https://user-images.githubusercontent.com/8896153/234409312-f8e1dd81-d016-4a53-9631-64f88e45574a.png">

---
id: validators
title: Input & Output Validators
sidebar_label: Input & Output Validators
slug: /server/validators
---

> tRPC procedures may define an input and/or an output. It's flexible too, so we support many different validators, and it's easy to integrate a validator which we don't directly support.

Reviewer 2: ```suggestion
tRPC procedures may define validation logic for their input and/or output. There is first class support for many different validators, and it's easy to integrate a validator which we don't directly support.
```

export const t = initTRPC.create();

// ---cut---

const procedureWhichRejectsOutOfTownFolk = t.procedure
  .input(z.object({ townName: z.string() }))
  .use((opts) => {
    if (opts.input.townName !== 'Pucklechurch') {
      throw new TRPCError({
        code: 'FORBIDDEN',
        message: "We don't take kindly to out-of-town folk",
      });
    }

    return opts.next();
  });

export const appRouter = t.router({
  hello: procedureWhichRejectsOutOfTownFolk
    .input(
      z.object({
        name: z.string(),
      }),
    )
    .query((opts) => {
      const input = opts.input;
      //    ^?
      return {
        greeting: `Hello ${input.name}, my friend`,
      };
    }),
});
```

### Output Validators

By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.

Validators are used to infer the input and output types of your procedures.

:::info
If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
:::

> ## Using a function as a validator

Reviewer 2: maybe this section makes more sense after the integrations?

Author: I think I'm trying to angle it as an initial representation of what a validator is doing. I'll do another pass now and see if you feel the same (happy to move it, I considered both ways as a coin toss)

---
id: validators
title: Input & Output Validators
sidebar_label: Input & Output Validators
slug: /server/validators
---

> tRPC procedures may define validation logic for their input and/or output. There is first class support for many popular validators, and [integrating a validator](#contributing-your-own-validator-library) which we don't directly support in usually straight-forward.

Reviewer 2: ```suggestion
tRPC procedures may define validation logic for their input and/or output. There is first class support for many popular validators, and you can also [integrate validators](#contributing-your-own-validator-library) which we don't directly support.
```

    )
    .query((opts) => {
      const name = opts.input.name;
      //    ^?
      return {
        greeting: `Hello ${opts.input.name}`,
      };
    }),
});
```

`.input` can also be stacked to build more complex types, which is particularly useful when you want to utilise some common input to a collection of procedures in a [middleware](middlewares).

```ts twoslash
// @target: esnext
import { TRPCError, initTRPC } from '@trpc/server';
import { z } from 'zod';

export const t = initTRPC.create();

// ---cut---

const procedureWhichRejectsOutOfTownFolk = t.procedure
  .input(z.object({ townName: z.string() }))
  .use((opts) => {
    if (opts.input.townName !== 'Pucklechurch') {
      throw new TRPCError({
        code: 'FORBIDDEN',
        message: "We don't take kindly to out-of-town folk",
      });
    }

    return opts.next();
  });

export const appRouter = t.router({
  hello: procedureWhichRejectsOutOfTownFolk
    .input(
      z.object({
        name: z.string(),
      }),
    )
    .query((opts) => {
      const input = opts.input;
      //    ^?
      return {
        greeting: `Hello ${input.name}, my friend`,
      };
    }),
});
```

### Output Validators

> By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.

Reviewer 2: ```suggestion
You can also define an output validator. Validating outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures. Some reasons to define an output validator include:

- Checking that data returned from untrusted sources is correct
- Ensure that you are not returning more data to the client than necessary
```

> Validators are used to infer the input and output types of your procedures.

Reviewer 2: does this belong here, or maybe further up in the general validation bit?

:::info
If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
:::

## The most basic validator: a function

> You can define a validator without any 3rd party dependencies, with a function. This isn't necessarily the best choice to use widely, unless you have a specific need, but it's important to understand that there's no magic here - it's *just typescript*!

Reviewer 2: ```suggestion
You can define a validator without any 3rd party dependencies, with a function. We don't recommend doing this unless you have a specific need, but it's important to understand that there's no magic here - it's *just typescript*!
```

Maybe also mention the "spec" for validators?
edit: saw the link to the interfaces below, but some reference to what a validator should do (return stuff or throw, anything else?) might make sense here
